pkg/controller/storagecluster/cephcluster.go

```diff
@@ -366,24 +403,61 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster) []rook.StorageClassDevi
 
 		for i := 0; i < replica; i++ {
 			placement := rook.Placement{}
+			preparePlacement := rook.Placement{}
 
 			if noPlacement {
-				in := getPlacement(sc, "osd")
-				(&in).DeepCopyInto(&placement)
-
-				if len(topologyKeyValues) >= replica {
-					podAffinityTerms := placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution
-					podAffinityTerms[0].PodAffinityTerm.TopologyKey = topologyKey
-
-					topologyIndex := i % len(topologyKeyValues)
-					nodeZoneSelector := corev1.NodeSelectorRequirement{
-						Key:      topologyKey,
-						Operator: corev1.NodeSelectorOpIn,
-						Values:   []string{topologyKeyValues[topologyIndex]},
+				if supportTSC {
+					in := getPlacement(sc, "osd-tsc")
+					(&in).DeepCopyInto(&placement)
+				} else {
+					in := getPlacement(sc, "osd")
+					(&in).DeepCopyInto(&placement)
+
+					if len(topologyKeyValues) >= replica {
+						placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution[0].PodAffinityTerm.TopologyKey = topologyKey
+
+						topologyIndex := i % len(topologyKeyValues)
+						nodeZoneSelector := corev1.NodeSelectorRequirement{
+							Key:      topologyKey,
+							Operator: corev1.NodeSelectorOpIn,
+							Values:   []string{topologyKeyValues[topologyIndex]},
+						}
+						appendNodeRequirements(&placement, nodeZoneSelector)
 					}
-					appendNodeRequirements(&placement, nodeZoneSelector)
 				}
+
+				if !noPreparePlacement {
```

Comment:
Lines 409-427 and lines 429-455 are very similar. Can you refactor to a helper method to avoid duplication of the pattern?

Reply:
Refactored the setting of topology key/values for affinity placements to a helper method and modified the `if structure` to avoid code duplication as mush as possible.